File: 1/ruby/part2.rb
# frozen_string_literal: true

WORD_TO_DIGIT = {
  'one' => '1',
  'two' => '2',
  'three' => '3',
  'four' => '4',
  'five' => '5',
  'six' => '6',
  'seven' => '7',
  'eight' => '8',
  'nine' => '9',
  'zero' => '0',
}.freeze

REVERSE_WORD_TO_DIGIT = WORD_TO_DIGIT.to_h { |k, v| [k.reverse, v] }

def calibration_sum(document)
  document.lines.sum { |line| Integer(first_digit(line) + last_digit(line)) }
end

def first_digit(line)
  alphabetic_substring = ''
  line.each_char do |char|
    return char if char.ord >= '0'.ord && char.ord <= '9'.ord

    alphabetic_substring += char
    WORD_TO_DIGIT.each do |number_word, digit|
      return digit if alphabetic_substring.end_with?(number_word)
    end
  end
end

def last_digit(line)
  alphabetic_substring = ''
  line.reverse.each_char do |char|
    return char if char.ord >= '0'.ord && char.ord <= '9'.ord

    alphabetic_substring += char
    REVERSE_WORD_TO_DIGIT.each do |number_word, digit|
      return digit if alphabetic_substring.end_with?(number_word)
    end
  end
end

document = File.read('../input')
puts calibration_sum(document)

File: 1/rust/src/part2.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::iter::Iterator;

const WORD_TO_DIGIT: [(&str, u32); 10] = [
    ("one", 1),
    ("two", 2),
    ("three", 3),
    ("four", 4),
    ("five", 5),
    ("six", 6),
    ("seven", 7),
    ("eight", 8),
    ("nine", 9),
    ("zero", 0),
];

const REVERSE_WORD_TO_DIGIT: [(&str, u32); 10] = [
    ("eno", 1),
    ("owt", 2),
    ("eerht", 3),
    ("ruof", 4),
    ("evif", 5),
    ("xis", 6),
    ("neves", 7),
    ("thgie", 8),
    ("enin", 9),
    ("orez", 0),
];

pub fn run() {
    let input_file = File::open("../input").unwrap();
    let lines = io::BufReader::new(input_file);
    let sum: u32 = lines.lines()
        .filter_map( |line| Some(line_value(line.ok()?)) )
        .map( |v| v.unwrap() )
        .sum();
    println!("{}", sum)
}

fn line_value(line: String) -> Option<u32> {
    Some(10 * first_digit(&line)? + last_digit(&line)?)
}

fn first_digit(line: &str) -> Option<u32> {
    let mut alphabetic_substring = String::from("");
    for c in line.chars() {
        if c.is_ascii_digit() {
            return c.to_digit(10)
        }

        alphabetic_substring.push(c);
        for (word, digit) in WORD_TO_DIGIT {
            if alphabetic_substring.ends_with(word) {
                return Some(digit)
            }
        }
    }

    None
}

fn last_digit(line: &str) -> Option<u32> {
    let mut alphabetic_substring = String::from("");
    for c in line.chars().rev() {
        if c.is_ascii_digit() {
            return c.to_digit(10)
        }

        alphabetic_substring.push(c);
        for (word, digit) in REVERSE_WORD_TO_DIGIT {
            if alphabetic_substring.ends_with(word) {
                return Some(digit)
            }
        }
    }

    None
}
